// @tanstack/router-plugin/dist/esm/core/router-generator-plugin.js
import { join, isAbsolute, normalize } from "node:path";
import { Generator, resolveConfigPath } from "@tanstack/router-generator";
import { getConfig } from "./config.js";
const PLUGIN_NAME = "unplugin:router-generator";
const unpluginRouterGeneratorFactory = (options = {}) => {
  const ROOT = process.cwd();
  let userConfig = options;
  let generator;
  const routeGenerationDisabled = () => userConfig.enableRouteGeneration === false;
  const getRoutesDirectoryPath = () => {
    return isAbsolute(userConfig.routesDirectory) ? userConfig.routesDirectory : join(ROOT, userConfig.routesDirectory);
  };
  const initConfigAndGenerator = () => {
    userConfig = getConfig(options, ROOT);
    generator = new Generator({
      config: userConfig,
      root: ROOT
    });
  };
  const generate = async () => {
    if (routeGenerationDisabled()) {
      return;
    }
    try {
      await generator.run();
    } catch (e) {
      console.error(e);
    }
  };
  const handleFile = async (file, event) => {
    const filePath = normalize(file);
    if (filePath === resolveConfigPath({ configDirectory: ROOT })) {
      initConfigAndGenerator();
      return;
    }
    try {
      await generator.run({ path: filePath, type: event });
    } catch (e) {
      console.error(e);
    }
  };
  return {
    name: "router-generator-plugin",
    enforce: "pre",
    async watchChange(id, { event }) {
      if (!routeGenerationDisabled()) {
        await handleFile(id, event);
      }
    },
    async buildStart() {
      await generate();
    },
    vite: {
      configResolved() {
        initConfigAndGenerator();
      },
      async buildStart() {
        if (this.environment.config.consumer === "server") {
          return;
        }
        await generate();
      },
      sharedDuringBuild: true
    },
    rspack(compiler) {
      initConfigAndGenerator();
      let handle = null;
      compiler.hooks.beforeRun.tapPromise(PLUGIN_NAME, generate);
      compiler.hooks.watchRun.tapPromise(PLUGIN_NAME, async () => {
        if (handle) {
          return;
        }
        const routesDirectoryPath = getRoutesDirectoryPath();
        const chokidar = await import("chokidar");
        handle = chokidar.watch(routesDirectoryPath, { ignoreInitial: true }).on("add", generate);
        await generate();
      });
      compiler.hooks.watchClose.tap(PLUGIN_NAME, async () => {
        if (handle) {
          await handle.close();
        }
      });
    },
    webpack(compiler) {
      userConfig = getConfig(options, ROOT);
      let handle = null;
      compiler.hooks.beforeRun.tapPromise(PLUGIN_NAME, generate);
      compiler.hooks.watchRun.tapPromise(PLUGIN_NAME, async () => {
        if (handle) {
          return;
        }
        const routesDirectoryPath = getRoutesDirectoryPath();
        const chokidar = await import("chokidar");
        handle = chokidar.watch(routesDirectoryPath, { ignoreInitial: true }).on("add", generate);
        await generate();
      });
      compiler.hooks.watchClose.tap(PLUGIN_NAME, async () => {
        if (handle) {
          await handle.close();
        }
      });
      compiler.hooks.done.tap(PLUGIN_NAME, () => {
        console.info("✅ " + PLUGIN_NAME + ": route-tree generation done");
      });
    }
  };
};
export {
  unpluginRouterGeneratorFactory
};
